### PR TITLE
[FR] [DAC] further decouple reliance on default rule dir locations

### PR DESCRIPTION
**Pull request by eric-forte-elastic**
## Issues
https://github.com/elastic/detection-rules/issues/3619

## Summary

The primary purpose of this PR is to remove the hard coding of rules directories and replace them with the directories loaded from a config file. Either the default one, or a user generated one. 

Additionally, there are a few other considerations that are related to the implementation:
- This PR should make BBR rules directories optional
- In the config, we should have a separate path list for BBR rules dirs as they are handled differently
- This PR updates the current loading of rules to support a list of paths to be loaded, instead of one path. 
- This PR updates the example config generation to match the formatting specified from the above considerations. 

Open Design Quetsions:
- In the config, should also have an RTA section (expected to be optional as well)?
   - Currently this is not done, but maybe it should be?
- Need to decide whether or not deprecated rules are optional and whether or not we should make this be a configurable directory location (as opposed to `_deprecated` relative to the rules directory.)
    - Current decision is to store deprecated rules in `_deprecated` folder relative to rules directory.  

## Testing

To test one will need to run the following commands with both the default config and a custom config. 

- `make build`
- `make test-cli`
- `make test-remote-cli`

### Testing Results

Custom config

<details><summary>make</summary>
<p>

```shell

```
[make_build.txt](https://github.com/elastic/detection-rules/files/15308572/make_build.txt)

</p>
</details> 

<details><summary>make test-cli</summary>
<p>

```shell

```
[make_test_cli.txt](https://github.com/elastic/detection-rules/files/15308620/make_test_cli.txt)

</p>
</details> 

<details><summary>make test-remote-cli</summary>
<p>

```shell
detection-rules on  3619-frdac-further-decouple-reliance-on-default-rule-dir-locations [$✘!?] is  v0.1.0 via  v3.12.3 (detection-rules-build) on  eric.forte
❯ make test-remote-cli
Executing test_remote_cli script...
Running detection-rules remote CLI tests...
Performing a quick rule alerts search...
Requires .detection-rules-cfg.json credentials file set.
Loaded config file: /home/forteea1/Code/clean_mains/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

kibana_user: eric.forte
kibana_password:
No alerts detected
Performing a rule export...
Loaded config file: /home/forteea1/Code/clean_mains/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

kibana_user: eric.forte
kibana_password:
- skipping First Time Seen AWS Secret Value Accessed in Secrets Manager - ValidationError
3 rules exported
2 rules converted
2 saved to tmp-export
1 errors saved to tmp-export/_errors.txt
adfind_command_activity.toml  clearing_windows_console_history.toml  _errors.txt
Removing generated files...
Detection-rules CLI tests completed!

```

</p>
</details>

### CODE REVIEW

File: detection_rules/cli_utils.py
Comment (traut) [inline detection_rules/cli_utils.py:23]: a side thought -- @Mikaayenson @eric-forte-elastic how do you guys feel about going `black`-enforced styling? I see we're using flake8 linting but I think it might be beneficial to add at least `black` in the mix. TDS pipeline [runs](https://buildkite.com/elastic/protections-cloud/builds/15030#018b3f9a-1b49-4613-bb48-0c75caef334e) isort + flake8 + black + pyright, but it might be an overkill for us.

Comment (eric-forte-elastic) [reply detection_rules/cli_utils.py:23]: Generally speaking, I think we do want black in detection rules. We use black in most of our other repos, it would make sense to also do that here. But definitely something I think we would want to wait until DAC is merged back to main, or do it in main and merge into DAC, as it would make a lot of changes in many of the files.

Comment (Mikaayenson) [reply detection_rules/cli_utils.py:23]: ++ we also started exploring ruff in our other repos.

File: detection_rules/cli_utils.py
Comment (Mikaayenson) [inline detection_rules/cli_utils.py:189]: Might was well update the path below to use `Path`

File: detection_rules/navigator.py
Comment (Mikaayenson) [inline detection_rules/navigator.py:168]: Same logic in packaging.py. Maybe we can consolidate to a rule method?

Comment (eric-forte-elastic) [reply detection_rules/navigator.py:168]: Agreed :+1:

File: detection_rules/cli_utils.py
Comment (Mikaayenson) [reply detection_rules/cli_utils.py:189]: Based on what we decided. Might make more sense to do some of this in https://github.com/elastic/detection-rules/pull/3659

File: detection_rules/navigator.py
Comment (eric-forte-elastic) [reply detection_rules/navigator.py:168]: Added a function in TOMLRule called `get_rules_dir_path` I did not make it a property due to the `path` field in TOMLRule being optional, but could make it a property if prefered.

File: detection_rules/devtools.py
Comment (Mikaayenson) [inline detection_rules/devtools.py:735]: Should be the correct deprecated path related to the core rules folder.

File: detection_rules/rule.py
Comment (Mikaayenson) [inline detection_rules/rule.py:1331]: is_relative_to

File: tests/test_all_rules.py
Comment (Mikaayenson) [inline tests/test_all_rules.py:558]: is_relative_to ?

**Review by Mikaayenson — APPROVED** [review]
~~LGTM!~~

I think we should run the `make test-cli`. there may be a few gaps still missing.

File: tests/test_all_rules.py
Comment (eric-forte-elastic) [reply tests/test_all_rules.py:558]: For this one we want to leave it as it is part of the check

File: detection_rules/rule.py
Comment (eric-forte-elastic) [reply detection_rules/rule.py]: Updated

File: detection_rules/devtools.py
Comment (eric-forte-elastic) [reply detection_rules/devtools.py]: Updated

File: detection_rules/cli_utils.py
Comment (eric-forte-elastic) [reply detection_rules/cli_utils.py:189]: Agreed.

File: detection_rules/cli_utils.py
Comment (eric-forte-elastic) [reply detection_rules/cli_utils.py:23]: We should look at this in a different issue.

**Comment by Mikaayenson** [comment]
We should probably update the custom_rules `setup-config` to match the bbr rule changes.

File: detection_rules/main.py
Comment (Mikaayenson) [inline detection_rules/main.py:37]: This will impact things like import-rules-to-repo

**Comment by eric-forte-elastic** [comment]
> ~LGTM!~
> 
> I think we should run the `make test-cli`. there may be a few gaps still missing.

Agreed, I think it passed for me due to a specific config setup in for my environment. Will take a look :eyes:

File: detection_rules/main.py
Comment (eric-forte-elastic) [reply detection_rules/main.py:37]: Fixed

File: detection_rules/main.py
Comment (Mikaayenson) [inline detection_rules/main.py:122]: Do we need this? When would base_path be None since its generated from the rule name?

Comment (eric-forte-elastic) [reply detection_rules/main.py:122]: Good point, will remove

File: tests/test_transform_fields.py
Comment (Mikaayenson) [inline tests/test_transform_fields.py:88]:
```suggestion
        sample_rule = rc.load_dict(windows_rule, Path("privilege_escalation_test_windows.toml)))
```

Comment (eric-forte-elastic) [reply tests/test_transform_fields.py:88]: Path is optional, removed it as it does not correspond to any file path (given the user may not ever have a windows rule)

Comment (Mikaayenson) [reply tests/test_transform_fields.py:88]: Was mainly thinking about what we loose in terms of testing by changing from load_file to load_dict.

Comment (eric-forte-elastic) [reply tests/test_transform_fields.py:88]: Happy to add back if preferred.

Comment (Mikaayenson) [reply tests/test_transform_fields.py:88]: More or less something to think (nit) about.

File: CLI.md
Comment (Mikaayenson) [inline CLI.md:93]: This will give people the ability to place rules where they want which is nice!

File: detection_rules/mappings.py
Comment (Mikaayenson) [inline detection_rules/mappings.py:17]: Do we need this here?

File: detection_rules/rule_loader.py
Comment (Mikaayenson) [inline detection_rules/rule_loader.py:28]: I dont see this used in this file

File: detection_rules/mappings.py
Comment (eric-forte-elastic) [reply detection_rules/mappings.py]: No I do not think we do, removed :+1:

File: detection_rules/rule_loader.py
Comment (eric-forte-elastic) [reply detection_rules/rule_loader.py:28]: It is imported in other files like devtools.py and test_all_rules.py

Comment (eric-forte-elastic) [reply detection_rules/rule_loader.py:28]: It is also used on the following line (line 29)

Comment (Mikaayenson) [reply detection_rules/rule_loader.py:28]: lol yup for some reason I didn't see this before!

**Review by Mikaayenson — APPROVED** [review]
⭐ Great work. It lgtm. Especially after making the final changes based on the make test files.

**Review by traut — APPROVED** [review]
looks good!

**Pull request merged**
